[Ho Kai Ting] iP

kaitinghh opened the pull request:
## JerryBot

> A little mischief never hurt nobody. 

Too busy? JerryBot helps you keep track of all your todos, events and deadlines. It's
- easy to learn
- intuitive
- cute!

Here's what JerryBot can do for you!
- [x] Keeping track of tasks
- [x] Notify you of duplicate tasks
- [ ]  Reminding you of deadlines (coming up next!)
- [ ]  Managing a schedule (coming soon)

All you need to do is,
1. download it from [here](https://github.com/nus-cs2103-AY2324S2/ip/pull/451).
2. run it.
3. add your tasks.
4. let it manage your tasks for you 😉

### Note

Please check out the command list before starting out.

Commands to **get started**:

```
todo Read a book
deadline Complete project /by 2024-03-01
event Team meeting /from 2024-02-28 14:00 /to 2024-02-28 15:00
list
mark 2
delete 3
```

When you are done, simply send the command `bye`!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

kaitinghh committed:
Level 0. Rename, Greet, Exit

kaitinghh committed:
Level 1. Echo

kaitinghh committed:
Level 2. Add, List

kaitinghh committed:
Level 3. Mark as done

kaitinghh committed:
Level 4. ToDos, Events, Deadline

kaitinghh committed:
A-TextUiTesting

kaitinghh committed:
Level-5. Handle Errors

kaitinghh committed:
Level 6. Delete

kaitinghh committed:
Implement data storage

kaitinghh committed:
Merge branch 'branch-Level-7'

kaitinghh committed:
Convert datetime String to LocalDateTime

kaitinghh committed:
Merge branch 'branch-Level-8'

kaitinghh committed:
Rename dataStore class to Storage

kaitinghh committed:
Implement Parser, TaskList and Ui classes

kaitinghh committed:
Merge branch 'branch-A-MoreOOP'

kaitinghh committed:
Add exception and task packages

kaitinghh committed:
Merge branch 'branch-A-Packages'

kaitinghh committed:
Merge remote-tracking branch 'upstream/add-gradle-support' into branch-A-Gradle

kaitinghh committed:
Merge branch 'branch-A-Gradle'

kaitinghh committed:
Add JUnit tests for Parser and Task classes

kaitinghh committed:
Merge branch 'branch-A-JUnit'

Bandov reviewed (commented):
Code Review
General Impressions
Firstly, I would like to commend the effort and quality put into the development of the project. The code is well-structured, readable, and demonstrates a good understanding of the problem domain. It's evident that a significant amount of work has been dedicated to ensuring the logic is correct and efficient, which is greatly appreciated.

Strengths
Readability: The code is very readable, with clear naming conventions and concise functions that effectively encapsulate their intended functionality.
Efficiency: The logic of the main class is well structured and easily understood with a proper taskflow during proper execution.

Areas for Improvement
While the codebase is well-written and effective, there are several areas where we can improve, especially considering the project's delay and the non-OOP structure.

Refactoring to OOP: Refactoring the codebase to use an Object-Oriented approach could enhance its scalability and maintainability. OOP can help encapsulate related functionalities into classes, making the code more modular, easier to understand, and adapt to future requirements. This transition can be gradual, starting with core components that would benefit most from encapsulation and inheritance.

Modularization: Even without a full transition to OOP, increasing modularity could improve the code's organization. Breaking down the code into more discrete, independent modules or functions can facilitate parallel development and make it easier to manage complexity as the project grows.

Automated Testing: Implementing automated tests, if not already in place, can significantly speed up the development cycle by catching bugs early and reducing manual testing time. This could help in catching up with the deadlines.

Conclusion
Overall, the codebase is in an excellent state. The recommendations provided aim to build on the solid foundation already established, with a focus on ensuring the project can scale and adapt to future needs while addressing the current challenge of being behind schedule. Transitioning towards more OOP practices and enhancing modularity will be key steps in this process.

Bandov commented on src/main/java/BotChat.java:
Perhaps splitting the logic of this code into two portions will make future debugging and issues easier in the future and make it less nested.

Bandov commented on src/main/java/TaskList.java:
good use of containerisation and data hiding techniques that I should follow!

Bandov commented on src/test/java/parser/ParserTest.java:
great use of tests and clear documentation for what is each test for

ashleyy2444 commented on src/main/java/BotChat.java:
```diff
@@ -0,0 +1,160 @@
+            }
+            return stringBuilder.toString();
+        } else if (markMatcher.matches()) {
+            try {
```
To improve readability, I think you can create methods for each case instead of writing the code here.

kaitinghh replied:
Thanks for pointing that out! I have abstracted out the functionalities of the code to separate methods.

ashleyy2444 commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,170 @@
+import task.Task;
+import task.ToDo;
+
+import java.nio.file.*;
```
The ordering of your import statements can be more consistent (eg. label all java.io first, followed by java.nio and then java.util).

ashleyy2444 commented on src/test/java/parser/ParserTest.java:
```diff
@@ -0,0 +1,40 @@
+
+public class ParserTest {
+    @Test
+    public void extractCommand_twoWordInput_returnsExtractedCommand() {
```
Can try to shorten all the method names as they are slightly lengthy.

ashleyy2444 commented on src/main/java/BotChat.java:
```diff
@@ -0,0 +1,160 @@
+    private static Storage storage;
+    private static TaskList taskArrayList;
+    private static Parser parser;
+    private static Pattern markPattern = Pattern.compile("mark \\d+");
```
Good use of Pattern class!

ashleyy2444 reviewed (commented):
Overall, the code was very interesting to read! It is good that you put all the relevant classes into different packages, making it more readable and easy to understand the code structure. However, I noticed that some methods implemented were quite long. I think you can try to break the longer methods down by creating more shorter methods so that it can increase readability. 
LGTM. Just a few nits to fix.

triciiaaa commented on src/main/java/BotChat.java:
```diff
@@ -0,0 +1,160 @@
+                break;
+            default:
+                throw new InvalidCommandException(s);
+        }
```
Should there be indentation for the `case` and `default` statement?

triciiaaa commented on src/main/java/task/Event.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+public class Event extends Task {
```
Would it be better to add a line break in between to enhance readibility?

triciiaaa commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,170 @@
+import task.ToDo;
+
+import java.nio.file.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.io.IOException;
+import java.io.BufferedWriter;
+import java.io.BufferedReader;
+import java.io.FileWriter;
+import java.io.FileReader;
+import java.util.Objects;
+import java.nio.file.Files;
```
Perhaps you can consider arranging the import statements in lexicographical order and add line breaks between them appropriately?

triciiaaa commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,34 @@
+    public static final String HI_MESSAGE = "Hello! I'm BotChat.\n What can I do for you?";
+    public static final String BYE_MESSAGE = "Bye. Hope to see you again soon!";
+
+    public static String hiMessage() {
+        return HI_MESSAGE;
+    }
+
+    public static String byeMessage() {
+        return BYE_MESSAGE;
+    }
```
Would it be better to rename the methods as verbs, i.e. `printHiMessage()` and `printByeMessage()` respectively?

kaitinghh replied:
Thank you for your review. I have edited my code to enhance readability and adhere to the coding standards as you have suggested.

triciiaaa reviewed (commented):
Overall, I found your code easy to read for the most part and I like how you used the Pattern class in your code! Good job 👍

kaitinghh committed:
Add JavaDoc for Parser, BotChat, Storage and TaskList classes

kaitinghh committed:
Merge branch 'branch-A-JavaDoc'

kaitinghh committed:
Rename boolean variables to follow coding standards

kaitinghh committed:
Break line to follow coding standards

kaitinghh committed:
Add find task by keyword functionality

kaitinghh committed:
Merge branch 'branch-A-CodingStandard'

kaitinghh committed:
Merge branch-Level-9 into master

kaitinghh committed:
Add Checkstyle

kaitinghh committed:
Merge branch 'branch-A-CheckStyle'

kaitinghh committed:
Add basic GUI

kaitinghh committed:
Change GUI to use FXML

kaitinghh committed:
Merge branch 'branch-Level-10'

kaitinghh committed:
Rename bot to JerryBot

kaitinghh committed:
Add assertions to Storage and TaskList classes

Storage and TaskList classes do not have assertions.

Storage and TaskList classes require non-null inputs to
initialise.

Let's use assertions to enforce these invariants of
non-null inputs for initialisation.

kaitinghh committed:
Add assertions to Storage and TaskList classes

Storage and TaskList classes do not have assertions.

Storage and TaskList classes require non-null inputs to initialise.

Let's use assertions to enforce these invariants of non-null inputs for initialisation.

kaitinghh committed:
Edit imports to improve code quality

Use wildcard imports.

We want to be specific with the packages we are
importing to adhere to code quality.

Let's import the specific packages that we require
in the class.

kaitinghh committed:
Remove unused imports and fix checkstyles

There are unused imports and checkstyle errors.

Unused imports make the code unnecessarily cluttered and reduces code
quality. Checkstyle errors make the code less readable.

Let's improve code quality by removing unused imports and correcting
checkstyle errors.

kaitinghh committed:
Merge pull request #2 from kaitinghh/branch-A-Assertions

Branch A-Assertions

kaitinghh committed:
Merge branch 'master' into branch-A-CodeQuality

kaitinghh committed:
Merge pull request #3 from kaitinghh/branch-A-CodeQuality

Branch A-CodeQuality

kaitinghh committed:
Prevent addition of duplicate tasks

kaitinghh committed:
Merge branch 'branch-BCD-Extension'

kaitinghh committed:
Merge remote-tracking branch 'origin'

kaitinghh committed:
Add Ui screenshot

kaitinghh committed:
Add user guide

kaitinghh committed:
Fix checkstyle violations

kaitinghh committed:
Update text ui testing

kaitinghh committed:
Abstract list and find messages out to Ui class

kaitinghh committed:
Edit package names to comply with style

kaitinghh committed:
Add Javadocs

kaitinghh committed:
Edit code to follow coding standards

kaitinghh committed:
Edit Ui method names to verbs

kaitinghh committed:
Change gradle configurations

kaitinghh committed:
Update UG & fix delete index bug

kaitinghh committed:
Update UG, fix mark & unmark index bugs
